### DISCUSSION TITLE
Obtain statements generated by a session.commit()

### DISCUSSION THREAD

**Discussion by Neah-Ko** [sqlalchemy/sqlalchemy#11068] [Usage Questions]
Hello,

I am originally looking for a generic way to create an entity with its eventual nested relationships as in #7336 

I am building a rest API leveraging `starlette` for asgi, `marshmallow` for validation and serialization and `SQLAlchemy` for ORM.

## Example
I am going for an architecture where server instance 'adopts' controllers (1 per entity) implementing HTTP methods. Each entity has its associated service class that performs db statements.

This example focuses on _Dataset_ creation. Its controller manages the route `POST /datasets/` redirecting to a function validating incoming data against marshmallow schema then sends down the resulting dict to the service.

- ORM snippet:
```python
class Dataset(Base):
    id:              Mapped[int] = Column(Integer, primary_key=True)
    name:            Mapped[str] = Column(String(50), nullable=False)
    id_user_contact: Mapped[int] = Column(ForeignKey("USER.id"), nullable=False)
    contact:         Mapped["User"]      = relationship(foreign_keys=[id_user_contact], lazy="immediate")
    tags:            Mapped[Set["Tag"]]  = relationship(secondary=asso_dataset_tag, lazy="immediate", uselist=True)

class User(Base):
    id = Column(Uuid, primary_key=True)

class Tag(Base):
    id = Column(Integer, nullable=False, primary_key=True)
    name = Column(String, nullable=False)
```

- POST request

The user nor tags exist, so we would like to create them in one go along with dataset item.

```bash
curl -d '{"name": "test", "contact": {"id": "613582fa-3be6-4954-bcc6-d2f1718ae826"}, "tags": [{"name":"bip"}, {"name":"bap"}, {"name":"bop"}]}' http://127.0.0.1:8000/datasets/
```

- DatasetSchema

```python
class DatasetSchema(Schema):
    id = Integer()
    version = Integer()
    name = String(required=True)
    id_user_contact = UUID(required=True,)
    contact = Nested('UserSchema', only=('id',))
    tags = List(Nested('TagSchema'), only=('name', ))

    @pre_load
    def pre_load_process(self, data, many, **kwargs):
        ret = data
        id_uc = data.get('id_user_contact')
        contact_id = data.get('contact', {}).get('id')
        if id_uc and contact_id:
            assert(id_uc == contact_id)
        elif contact_id and not id_uc:
            ret["id_user_contact"] = contact_id
        else:
            raise ValidationError("Need one of id_user_contact or contact fields.")
        return ret
```

- Marshmallow Schema validation result
```json
{
  "name": "ds_test",
  "id_user_contact": UUID('613582fa-3be6-4954-bcc6-d2f1718ae826'),
  "contact": {"id": UUID('613582fa-3be6-4954-bcc6-d2f1718ae826')}, 
  "tags": [{"name": "bip"}, {"name": "bap"}, {"name": "bop"}]
}
```

- Services snippet
```python
class DatabaseService(BaseService, metaclass=Singleton):
    async def _insert(self, stmt: Insert) -> (Any | None):
        async with self.db_session as s:
            row = await s.scalar(stmt)
        if row: return row

    async def _insert_many(self, stmt: Insert) -> List[Any]:
        async with self.db_session as s:
            rows = await s.scalars(stmt)
        return rows

class UnaryEntityService(DatabaseService):
    async def create(self, data, stmt_only: False) -> table | List[table]:
        if isinstance(data, list):
            stmt = insert(self.table).values(data).returning(self.table)
            return stmt if stmt_only else await self._insert_many(stmt)
        else:
            stmt = insert(self.table).values(**data).returning(self.table)
            return stmt if stmt_only else await self._insert(stmt)

class TagService(UnaryEntityService):
    pass

class UserService(UnaryEntityService):
    pass
```

My idea is to insert in bulk as much as possible in order to avoid session multiplication. So service/create methods are implementing a `stmt_only` flag in order to aggregate in the upper levels. Then the create for _composite_ entities is inspired by the discussion that I linked above:

```python
class CompositeEntityService(UnaryEntityService):
    async def _insert_composite(self, tpl: tuple[Insert, List[Insert]]) -> (Any | None):
        s_item, ls_nested = tpl
        async with self.db_session as s:
            # Insert all dependencies: important to await.
            for sub in ls_nested:
                await s.scalar(sub)
            # Insert item
            row = await s.scalar(s_item)
        if row: return row

    async def create(self, data, stmt_only=False) -> table:
        """CREATE, accounting for nested entitites."""
        stmts = []
        mapper = inspect(self.table).mapper
        rels = mapper.relationships
        delayed = {}

        # For all table relationships, check if data contains that item.
        for key in rels.keys():
            rel = rels[key]
            sub = data.get(key)
            if not sub: continue

            # Retrieve associated service.
            target_table = rel.target
            svc = target_table.name.capitalize() + "Service"
            svc = getattr(getattr(model, "services"), svc)()

            # Single nested entity: We build and add statement to batch.
            if isinstance(sub, dict):
                nested_stmt = await svc.create(sub, stmt_only=True)
                stmts += [nested_stmt]

            # List of entities: one - to - many relationship.
            # -> insert in a separate session: we need a handle on
            #    inserted items in order to populate the associative table.
            elif isinstance(sub, list):
                nested_items = await svc.create(sub, stmt_only=False)
                delayed[key] = nested_items

            # Remove from data dict to avoid statement errors.
            del data[key]

        # Statement for original item.
        stmt = insert(self.table).values(**data).returning(self.table)

        if stmt_only:
            # TODO: Handle that case
            raise NotImplementedError

        # Insert objects
        item = await self._insert_composite((stmt, stmts))

        # Populate lists and let session.merge insert associations.
        async with self.db_session as s:
            for key in delayed.keys():
                for e in delayed[key]:
                    getattr(item, key).add(e)
            await s.merge(item)
        return item

class DatasetService(CompositeEntityService):
    pass
```

## Issue

So this solution above is working neatly in the case I want to insert immediately.  Point being that the associative table of the relationship is 'automatically' filled by sqlalchemy when adding the inserted list of tags to the item, using `session.merge`  and committing.

However, to ensure genericity I would like to be able to Implement the path for `stmt_only` flag as well, so that a composite entity could be created from another one.

I've tried various approaches to have the session object return the statements it builds on commit E.g:

```python
async with self.db_session as s:
  trans = s.sync_session.get_transaction()
  trans.prepare()
  trans._prepare_impl()`
  subs = trans._iterate_self_and_parents(upto=trans)
  subs = [sub for sub in subs]
```

however a lot of things are going around with transactions, units of work and so on.. and the core is not happy about me tickling transaction objetcs.

It is getting difficult for me to analyze what is happening behind the hood. Intuitively I was looking for a way to have the session give me a handle on the statements it would issue on a `commit` without executing them.

Would you think that is possible ? Else what kind of approach would you recommend for this use case ?

Best.

- Packages:
```
starlette              0.36.3
SQLAlchemy             2.0.25
marshmallow            3.20.2
```

**Comment by zzzeek** [comment]
> My idea is to insert in bulk as much as possible in order to avoid session multiplication. 

This seems to be the crux of this whole thing,  are you "inserting in bulk" across many web requests, like after the fact?   or a single web request implies many INSERT statements?   

truth be told I so far stopped reading after that part since I dont understand what it is you're trying to do, I see things like "session.prepare()" and stuff like that and this is all way way in there and so far this smells like it's overengineered.  I need to understand the use case better:

1. one web request comes in?    What happens?  is a record INSERTed when the request returns?
2. are you using only session.execute(insert  statement) ?  or are you also trying to leverage unit of work flush features ?  (relationship resolution, etc).  these two approaches are kind of separate and dont really mix for a single entity

**Comment by Neah-Ko** [comment]
Hi @zzzeek ,

So I'm building a REST API, which accepts requests like `(GET | PUT | POST | ...) /{resource}s/{params}`. This resource can be a _composite entity_: contain nested entities.

In the case a resource creation is triggered (`POST -d {} /{resource}s/`), I wanted a way to have all generated `insert` statements execute in one session. Intuitively I would like that **one** request hitting the API leads to opening **one** asynchronous session. 

Originally I thought I could not generate associative table inserts (one-to-many relationship: nested list of entities) before the main item gets created and the db returns generated primary keys. I was using some _hacky_ workaround by merging the item in a new session then populate the list. 

In the end I went for a solution where  _hidden functions_ in the example above (`_select`, `_insert`, ...) may receive a session in parameter combined with the following decorator to ensures that it is set only once in a call stack.

```python
  def in_session(db_exec):
      """Decorator that ensures db_exec receives a session.

      contextlib.AsyncExitStack() below allows for conditional context management.
      """
      async def wrapper(self, stmt, session: AsyncSession=None):
          async with AsyncExitStack() as stack:
              if not session:
                  session = await stack.enter_async_context(self.app.db.session())
              return await db_exec(self, stmt, session)
      return wrapper
```

Then decorate:
```python
@in_session
    async def _select(...):
      ...

@in_session
    async def _insert(...

```

Then have a class `CompositeInsert` that acts as a statement and separates nested entities according to the relationship type (one-to-one, one-to-many) then an extra hidden function `_insert_composite(...)` that inserts nested entities (oto), then parent object, then populates the lists (o2m).

You may consult the codebase [here](https://github.com/bag-cnag/biodm/blob/main/src/model/dbservice.py) in case I failed in my explanations.

Now, is it **over-engineered** ? Well maybe. However, I did not found the section of the documentation for how to handle nested entities properly. I'm beginning using SQLAlchemy and have a more traditional DBMS background. I initially thought that given tables definition and a matching marshmallow schema I would be able to do this:


```python
class ResourceSchema(Schema):
   id = field.Integer()
   f1
   ...
  fn
```


```python
class ResourceTable(Base):
   id = Column(Integer, ...)
   f1
   ...
   fn
```

Then:
```bash
curl -d {"f1": "d1", ..., "fn": "dn"} HOST/resource/
```

and finally
```python
body = await request.body()
validated = ResourceSchema().loads(json_data=body)
stmt = insert(ResourceTable).values(**validated)
async with session as s:
   s.execute(stmt)
```

This works in the case of atomic fields but not for nested entities, nor nested list of entities. 
I need to instantiate the item first then populate attributes.

Anyway, I solved my problem. Thanks for your time.

Best,

**Comment by zzzeek** [comment]
> In the case a resource creation is triggered (POST -d {} /{resource}s/), I wanted a way to have all generated insert statements execute in one session. Intuitively I would like that one request hitting the API leads to opening one asynchronous session.

In my mind that's like the default way that any RESTful API would work;   the Session is set up front at the top of a web request, and all functions / methods called within the request make use of that same session.  That is, this is literally the most simple case there is.  That's why this seems "over engineered" if you can pardon the term - if your system has some architecture built of many smaller methods all of which open and close their own Session, that would be the anti-pattern in play that is leading to all this complexity.  The section at https://docs.sqlalchemy.org/en/20/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it talks about this explicitly.    

> As a general rule, the application should manage the lifecycle of the session externally to functions that deal with specific data. This is a fundamental separation of concerns which keeps data-specific operations agnostic of the context in which they access and manipulate that data.
> Keep the lifecycle of the session (and usually the transaction) separate and external.

**Reply by Neah-Ko** [reply]
Yeah definitely, I realized this as well. 

I took inspiration from a project out there that was implementing this anti-pattern then realized it while documenting this discussion. 
I think my current implementation is fine, now logs are only showing one BEGIN/COMMIT per request.
Thanks.

**Reply by zzzeek** [reply]
one thing that may not be apparent about the Session is that calling begin()/commit() is IO-free.   IO with the database, and a real database transaction, only starts up once one is needed.   this is all designed with this use case in mind
